vault/identity_store_entities.go

```diff
@@ -0,0 +1,463 @@
+	}
+
+	return i.handleEntityUpdateCommon(req, d, entity)
+}
+
+// handleEntityUpdateCommon is used to update an entity
+func (i *identityStore) handleEntityUpdateCommon(req *logical.Request, d *framework.FieldData, entity *entityStorageEntry) (*logical.Response, error) {
+	var err error
+	var newEntity bool
+
+	// Entity will be nil when a new entity is being registered; create a new
+	// struct in that case.
+	if entity == nil {
+		entity = &entityStorageEntry{}
+		newEntity = true
+	}
+
+	// Update the policies
+	entity.Policies = d.Get("policies").([]string)
+
+	// Get the name
+	entityName := d.Get("name").(string)
+	if entityName != "" {
+		entityByName, err := i.memDBEntityByName(entityName)
+		if err != nil {
+			return nil, err
+		}
+		switch {
+		case (newEntity && entityByName != nil), (entityByName != nil && entity.ID != "" && entityByName.ID != entity.ID):
+			return logical.ErrorResponse("entity name is already in use"), nil
+		}
+		entity.Name = entityName
+	}
+
+	// Get entity metadata
+
+	// Accept metadata in the form of map[string]string to be able to index on
+	// it
+	var entityMetadata map[string]string
+	entityMetadataRaw, ok := d.GetOk("external_metadata")
+	if ok {
+		entityMetadataInput := entityMetadataRaw.(map[string]interface{})
+		if len(entityMetadataInput) > 0 {
+			entityMetadata = make(map[string]string, len(entityMetadataInput))
+			for k, v := range entityMetadataInput {
+				vStr, ok := v.(string)
+				if !ok {
+					return logical.ErrorResponse(fmt.Sprintf("invalid metadata value %#v for key %q", v, k)), nil
+				}
+				entityMetadata[k] = vStr
+			}
+		}
+	}
+
+	entity.ExternalMetadata = entityMetadata
+
+	// ID creation and some validations
+	err = sanitizeEntity(entity)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the identities input
+	var identitiesSlice []interface{}
+	identitiesRaw, ok := d.Raw["identities"]
+	if ok {
+		identitiesSlice = identitiesRaw.([]interface{})
+	}
+
+	// Prepare parsed collection of given identities
+	var identities []*identityIndexEntry
+	for _, identityRaw := range identitiesSlice {
+		var input identityInput
+		if err = mapstructure.WeakDecode(identityRaw, &input); err != nil {
+			return nil, fmt.Errorf("failed to decode identity input: %v", err)
+		}
+
+		if input.MountPath == "" {
+			return logical.ErrorResponse("missing mount path"), nil
+		}
+
+		mountValidationResp := i.validateMountPathFunc(input.MountPath)
+		if mountValidationResp == nil {
+			return logical.ErrorResponse(fmt.Sprintf("invalid mount path %q", input.MountPath)), nil
+		}
+
+		identity := &identityIndexEntry{
+			EntityID:         entity.ID,
+			Name:             input.Name,
+			ExternalMetadata: input.ExternalMetadata,
+			MountID:          mountValidationResp.MountID,
+			MountType:        mountValidationResp.MountType,
+		}
+
+		err = i.sanitizeIdentity(identity)
+		if err != nil {
+			return nil, err
+		}
+
+		identities = append(identities, identity)
```

Comment:
It doesn't seem like you're checking to see whether the given name/mount ID combo is already in use somewhere, or for that matter whether any of the identities being specified right here are duplicates of each other.

I also wonder about unilaterally replacing existing identity entries. Does this require updates to remove any identities not kept from the indexes? Does it seem too onerous to require the administrator to specify every entity every time? I realize it's the same boat as the metadata but it feels easy to footgun. Maybe in both cases we should allow updates? Although that would require having a decent mechanism for deletion.